File: bindings/csharp/FastParse/ParseOptions.cs
namespace FastParse;

/// <summary>
/// Options passed to a FastParse parse call.
/// </summary>
public sealed class ParseOptions
{
    /// <summary>
    /// Language grammar to use. The default package supports <c>java</c>; loaded extensions can add more languages.
    /// </summary>
    public string Language { get; init; } = "java";

    /// <summary>
    /// Output format to produce.
    /// </summary>
    public FastParseFormat Format { get; init; } = FastParseFormat.Json;

    /// <summary>
    /// Optional pipe-delimited Tree-sitter rule names to include, for example <c>class_declaration|method_declaration</c>.
    /// Leave null or empty to include the full AST.
    /// </summary>
    public string? IncludeRules { get; init; }

    /// <summary>
    /// Field flags to include in output. <see cref="FastParseField.DefaultAll"/> uses the native default.
    /// </summary>
    public FastParseField Fields { get; init; } = FastParseField.DefaultAll;

    /// <summary>
    /// Whether token nodes should be included when supported by the native parser.
    /// </summary>
    public bool IncludeTokens { get; init; }

    /// <summary>
    /// Whether JSON output should be pretty-printed.
    /// </summary>
    public bool Pretty { get; init; }

    /// <summary>
    /// Source normalization to apply before parsing. AutoSafe applies conservative legacy cleanup for languages such as COBOL.
    /// </summary>
    public FastParseNormalization Normalization { get; init; } = FastParseNormalization.AutoSafe;

    /// <summary>
    /// Default JSON parse options that return the exploratory AST output.
    /// </summary>
    public static ParseOptions JsonAll { get; } = new();
}

File: bindings/java/fastparse/src/main/java/dev/fastparse/ParseOptions.java
package dev.fastparse;

/** Immutable options passed to a FastParse parse call. Defaults to full Java JSON AST. */
public final class ParseOptions {
    private final String language;
    private final FastParseFormat format;
    private final String includeRules;
    private final long fields;
    private final boolean includeTokens;
    private final boolean pretty;
    private final FastParseNormalization normalization;

    private ParseOptions(Builder builder) {
        this.language = builder.language;
        this.format = builder.format;
        this.includeRules = builder.includeRules;
        this.fields = builder.fields;
        this.includeTokens = builder.includeTokens;
        this.pretty = builder.pretty;
        this.normalization = builder.normalization;
    }

    public static Builder builder() {
        return new Builder();
    }

    public String getLanguage() { return language; }
    public FastParseFormat getFormat() { return format; }
    public String getIncludeRules() { return includeRules; }
    public long getFields() { return fields; }
    public boolean isIncludeTokens() { return includeTokens; }
    public boolean isPretty() { return pretty; }
    public FastParseNormalization getNormalization() { return normalization; }

    public static final class Builder {
        private String language = "java";
        private FastParseFormat format = FastParseFormat.JSON;
        private String includeRules;
        private long fields = 0L;
        private boolean includeTokens;
        private boolean pretty;
        private FastParseNormalization normalization = FastParseNormalization.AUTO_SAFE;

        public Builder language(String language) {
            this.language = language == null || language.isEmpty() ? "java" : language;
            return this;
        }

        public Builder format(FastParseFormat format) {
            this.format = format == null ? FastParseFormat.JSON : format;
            return this;
        }

        /** Pipe-delimited exact Tree-sitter rule names. Null/empty means full AST. */
        public Builder includeRules(String includeRules) {
            this.includeRules = includeRules == null || includeRules.isEmpty() ? null : includeRules;
            return this;
        }

        /** Native field mask. Zero means native default/all fields. */
        public Builder fields(long fields) {
            this.fields = fields;
            return this;
        }

        public Builder fields(FastParseField... fields) {
            this.fields = FastParseField.maskOf(fields);
            return this;
        }

        public Builder includeTokens(boolean includeTokens) {
            this.includeTokens = includeTokens;
            return this;
        }

        public Builder pretty(boolean pretty) {
            this.pretty = pretty;
            return this;
        }

        public Builder normalization(FastParseNormalization normalization) {
            this.normalization = normalization == null ? FastParseNormalization.AUTO_SAFE : normalization;
            return this;
        }

        public ParseOptions build() {
            return new ParseOptions(this);
        }
    }
}
